packages/docusaurus-theme-mermaid/src/theme/useMermaid.tsx

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {useEffect} from 'react';
+import mermaid from 'mermaid';
+import useIsBrowser from '@docusaurus/useIsBrowser';
+import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
+import type mermaidAPI from 'mermaid/mermaidAPI';
+import type {ThemeConfig} from '@docusaurus/theme-mermaid';
+
+const DEFAULT_DARK_THEME = 'dark' as mermaidAPI.Theme.Dark;
+const DEFAULT_LIGHT_THEME = 'default' as mermaidAPI.Theme.Default;
+
+const DARK_THEME_KEY = 'dark';
+const LIGHT_THEME_KEY = 'light';
+
+const HTML_THEME_ATTRIBUTE = 'data-theme';
+
+/**
+ * Gets the theme based on config and current data-theme of the HTML.
+ *
+ * @param html The HTML element of the page.
+ * @param config The configuration for this chart.
+ */
+function getTheme(
+  html: HTMLHtmlElement,
+  options?: ThemeConfig['mermaid'],
+): mermaidAPI.Theme {
+  let htmlTheme = html.getAttribute(HTML_THEME_ATTRIBUTE) ?? LIGHT_THEME_KEY;
+
+  if (!(htmlTheme === LIGHT_THEME_KEY || htmlTheme === DARK_THEME_KEY)) {
+    htmlTheme = LIGHT_THEME_KEY;
+  }
+
+  const defaultTheme =
+    htmlTheme === LIGHT_THEME_KEY ? DEFAULT_LIGHT_THEME : DEFAULT_DARK_THEME;
+
+  return options?.theme?.[htmlTheme] ?? options?.config?.theme ?? defaultTheme;
+}
+
+export default function useMermaid(): void {
```

Comment:
Let's make this a client module of `theme-mermaid`, shall we? This way we don't need to occupy `Layout`.

Reply:
It didn't work as a client module as the `mermaid` variable was in a different scope to the diagrams. Added `useMermaid` to the component to ensure that the intialise is only called once when the theme changes

Reply:
This hook is pure side-effect, though; I don't see obvious reasons why it won't work 🤔 Maybe I'm overlooking something. I'll play with it later.

Reply:
This is what I had for it:

```ts
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import mermaid from 'mermaid';
import ExecutionEnvironment from '@docusaurus/ExecutionEnvironment';
import siteConfig from '@generated/docusaurus.config';
import type mermaidAPI from 'mermaid/mermaidAPI';
import type {ThemeConfig} from '@docusaurus/theme-mermaid';

const DEFAULT_DARK_THEME = 'dark' as mermaidAPI.Theme.Dark;
const DEFAULT_LIGHT_THEME = 'default' as mermaidAPI.Theme.Default;

const DARK_THEME_KEY = 'dark';
const LIGHT_THEME_KEY = 'light';

const HTML_THEME_ATTRIBUTE = 'data-theme';

/**
 * Gets the theme based on config and current data-theme of the HTML.
 *
 * @param html The HTML element of the page.
 * @param config The configuration for this chart.
 */
function getTheme(
  html: HTMLHtmlElement,
  options?: ThemeConfig['mermaid'],
): mermaidAPI.Theme {
  let htmlTheme = html.getAttribute(HTML_THEME_ATTRIBUTE) ?? LIGHT_THEME_KEY;

  if (!(htmlTheme === LIGHT_THEME_KEY || htmlTheme === DARK_THEME_KEY)) {
    htmlTheme = LIGHT_THEME_KEY;
  }

  const defaultTheme =
    htmlTheme === LIGHT_THEME_KEY ? DEFAULT_LIGHT_THEME : DEFAULT_DARK_THEME;

  return options?.theme?.[htmlTheme] ?? options?.config?.theme ?? defaultTheme;
}

if (ExecutionEnvironment.canUseDOM) {
  const themeConfig = siteConfig.themeConfig as ThemeConfig;
  // Watch for changes in theme in the HTML attribute `data-theme`.
  const html: HTMLHtmlElement = document.querySelector('html')!;
  const init = (target: HTMLHtmlElement): void => {
    const theme = getTheme(target, themeConfig.mermaid);

    if (themeConfig.mermaid?.config) {
      mermaid.initialize({
        startOnLoad: true,
        ...themeConfig.mermaid.config,
        theme,
      });
    } else {
      mermaid.initialize({startOnLoad: true, theme});
    }

    html.setAttribute('data-mermaid', theme);
  };

  const observer = new MutationObserver((mutations) => {
    for (const mutation of mutations) {
      if (
        mutation.type !== 'attributes' ||
        mutation.attributeName !== 'data-theme'
      ) {
        continue;
      }
      init(mutation.target as HTMLHtmlElement);
      break;
    }
  });

  init(html);

  observer.observe(html, {attributes: true});
}
```

Reply:
Yes, that's also what I had in mind! Does it work?

Reply:
The html attribute `data-mermaid` gets updated, the `Mermaid` component detects this and rerenders but the theme doesn't change. So it's something to do with the call to `mermaid.initialize` not changing the theme for the diagrams calls to `mermaid.render`

Reply:
I see, it's because the effect is not re-run when the components re-render. I'm good 👍

I think this should only need to run once globally, not once for every MDX component, right? I'd have to think about it.

Reply:
Yeah this effect only needs to run once so that it can add a listener to the `html` element for the docusaurus theme changing. The listener is only used to tell mermaid that the theme has changed. The `mermaid` variable is global so the call to `initialize` should only be done once when the theme changes (hence this not being in the `Mermaid` component). The `data-mermaid` attribute is then used to tell the `Mermaid` components to rerender without them having a race condition on the `data-theme` attribute.

Reply:
Have you considered using `useColorMode`? Observing `data-theme` seems like an overkill to me, since the mutation itself is fired from the color mode change.

Reply:
I hadn't, it would probably work. mdx-mermaid doesn't have a dependency on docusaurus so with these changes I had originally stuck with what I knew worked

Reply:
OK, do you want to try it, or should I do it?

Reply:
It will probably be Monday before I get chance to give it a go, so you're welcome to. Otherwise I'll give it a go then

Reply:
Sure, I hope I can spare some time over the weekend.